source/include/protocol/cs_CuckooFilterStructs.h

```diff
@@ -0,0 +1,37 @@
+/*
+ * Author: Crownstone Team
+ * Copyright: Crownstone (https://crownstone.rocks)
+ * Date: Mar 26, 2021
+ * License: LGPLv3+, Apache License 2.0, and/or MIT (triple-licensed)
+ */
+#pragma once
+
+#include <stddef.h>
+#include <cstdint>
+
+typedef const void* cuckoo_key_t;
+typedef uint16_t cuckoo_fingerprint_t;
+typedef uint8_t cuckoo_index_t;
+
+/**
+ * Representation of an object (O) in this filter comprises of a fingerprint (F)
+ * and a the bucket index (A) where this fingerprint is located. Each object
+ * has an alternative location (B). Together (C):= (F,A,B) contains all the information
+ * that the filter retains about the original object (O). This data chunk (C) can be
+ * reused in several functions and thus got its own struct defintion.
+ */
+struct __attribute__((__packed__)) cuckoo_extended_fingerprint_t {
+	cuckoo_fingerprint_t fingerprint;
+	cuckoo_index_t bucketA;  // the bucket this fingerprint should be put in
+	cuckoo_index_t bucketB;  // and the alternative bucket.
+};
+
+/**
+ * Data content of the cuckoo filter.
+ */
+struct __attribute__((__packed__)) cuckoo_filter_data_t {
+	cuckoo_index_t bucket_count_log2; // we save the log2 value since bucket_count needs to be a power of 2.
+	cuckoo_index_t nests_per_bucket;
+	cuckoo_extended_fingerprint_t victim;
+	cuckoo_fingerprint_t bucket_array[];  // 'flexible array member'
```

Comment:
what is this array?

Reply:
This is a 'flexible array member' as described here https://en.cppreference.com/w/c/language/struct . It makes it possible for the struct to contain a data array of variable length without the overhead of an extra pointer.

E.g. if you change the type of `bycket_array` to `cuckoo_fingerprint_t*` the pointer takes up 4 bytes and the array needs to be allocated on top of that.